File: lib/javascript/fibonacci.js
// imports
const fs = require('fs');
const path = require('path');
const { writeToFileAsJson, checkIfFileExists } = require('./utils');

// constants
const maxFibNumber = 1474;
const args = process.argv.slice(2);
const secondsInMilliseconds = 1000;
const maxFibNumberForBasicRecursion = 45;
const dataFolder = path.join('data', 'fibonacci', 'javascript');

// fibonacci sequence algorithms
const fibIterative = n => {
    if (n < 2) return n;

    let a = 0, b = 1, f = 1;
    for (let i = 2; i <= n; i++) {
        f = a + b;
        a = b;
        b = f;
    }
    return f;
};

const fibRecursive = n => {
    if (n < 2) return n;
    return fibRecursive(n - 1) + fibRecursive(n - 2);
};

const fibRecursiveMemoized = (n, memo = {}) => {
    if (n < 2) return n;
    if (memo[n]) return memo[n];
    memo[n] = fibRecursiveMemoized(n - 1, memo) + fibRecursiveMemoized(n - 2, memo);
    return memo[n];
};

const fibBinetsFormula = n => {
    const SQRT_FIVE = 2.23606797749979;
    // (1 + SQRT_FIVE) / 2
    const FIRST_TERM = 1.618033988749895;
    // (1 - SQRT_FIVE) / 2
    const SECOND_TERM = -0.6180339887498949;

    if (n < 2) return n;

    return Math.round(((Math.pow(FIRST_TERM, n) - Math.pow(SECOND_TERM, n)) / SQRT_FIVE));
};

function calculate(_function, n) {
    const startTime = performance.now();
    const result = _function(n);
    const endTime = performance.now();
    const timeInSeconds = (startTime && endTime) ?
        (endTime - startTime) / secondsInMilliseconds : null;

    return { result, timeInSeconds };
}

function evaluate(_function, n) {
    return _function.name === 'fibRecursive' ?
        n <= maxFibNumberForBasicRecursion && calculate(_function, n) :
        calculate(_function, n);
}

// finds the numberInSequence for each algorithm and writes the results to a file
// not exported because the python scripts will have to be run separately anyways
function main(numberInSequence) {
    const results = {};
    const functions = [fibIterative, fibRecursive, fibBinetsFormula, fibRecursiveMemoized];
    const fileName = path.join(process.cwd(), dataFolder, `fib-js-${numberInSequence}.json`);

    numberInSequence = numberInSequence || maxFibNumber;

    if (!checkIfFileExists(fileName)) {
        for (const _function of functions) {// NOSONAR
            for (let n = 0; n < numberInSequence; n++) {
                const data = evaluate(_function, n);

                results[_function.name] = results[_function.name] || [];

                data && results[_function.name]
                    .push(data);
            }
        }
        // write the results to a file
        writeToFileAsJson(results, fileName);
    }
}


if (args.length > 0) {
    main(args[0]);
} else {
    module.exports = {
        /**
         * Finds the nth number in the Fibonacci Sequence using Binet's Formula
         *  @param {number} nthNumber - The nth number in the sequence
         * @return {BigInt|Infinity}  BigInt if calculation doesn't overflow, Infinity otherwise.
         *  Overflow will occur when n > 1474.
         */
        fibonacci: nthNum => fibBinetsFormula(nthNum),
        fibIterative,
        fibRecursive,
        fibBinetsFormula,
        fibRecursiveMemoized
    };
}

File: lib/python/fibonacci.py
import os
import sys
import json
import time

import utils.files as utils

max_fib_number = 1474
sys.setrecursionlimit(1500)
max_fib_num_for_basic_recursion = 40


def fib_recursive(n):
    """
    Returns the nth Fibonacci number, via recursion.
    """
    if n < 2:
        return n
    else:
        return fib_recursive(n-1) + fib_recursive(n-2)


def fib_recursive_memoized(n, memo={}):
    """
    Returns the nth Fibonacci number, via memoized recursion.
    """

    if n < 2:
        return n
    elif n in memo:
        return memo[n]
    else:
        memo[n] = fib_recursive_memoized(
            n-1, memo) + fib_recursive_memoized(n-2, memo)
        return memo[n]


def fib_iterative(n):
    """
    Returns the nth Fibonacci number, using an iterative approach.
    """
    if n < 2:
        return n
    else:
        a = 0
        b = 1
        for _ in range(2, n+1):
            c = a + b
            a = b
            b = c
        return c


def fib_binets_formula(n):
    """
    Returns the nth Fibonacci number, Using Binet's Formula.
    """
    sqrt_five = 2.23606797749979
    # 1 + sqrt(5) / 2
    first_term = 1.618033988749895
    # 1 - sqrt(5) / 2
    second_term = -0.6180339887498949

    return (1/sqrt_five)*((first_term)**n - (second_term)**n)


def calculate(func, current_fib_number):
    start_time = time.perf_counter()
    result = func(current_fib_number)
    end_time = time.perf_counter()

    time_in_seconds = end_time - start_time

    return {"result": result, "timeInSeconds": time_in_seconds}


def evaluate(func, current_fib_number):
    function_name = str(func.__name__)

    if function_name == 'fib_recursive':
        if current_fib_number <= max_fib_num_for_basic_recursion:
            return calculate(func, current_fib_number)
    else:
        return calculate(func, current_fib_number)


def main(args=None):  # NOSONAR
    results = {}
    functions = [
        fib_iterative,
        fib_recursive,
        fib_binets_formula,
        fib_recursive_memoized
    ]
    number_in_sequence = args if args else max_fib_number
    file_name = os.path.join(
        os.getcwd(), 'data', 'fibonacci', 'python', f"fib-py-{number_in_sequence}.json")

    if not utils.check_if_file_exists(file_name):
        for fun in functions:
            function_name = str(fun.__name__)

            if function_name not in results:
                results[function_name] = []

            for i in range(0, number_in_sequence):
                results[function_name] = results[function_name]\
                    if results[function_name] else []

                data = evaluate(fun, i)

                if data:
                    results[function_name].append(data)

        # Write to file
        return utils.write_to_file(file_name, json.dumps(results))


if __name__ == "__main__":
    _args = sys.argv[1:]

    try:
        main(int(_args[0]))
    except IndexError as e:
        main()
